onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+  if (node->OpType() == "Cast") {
+    const NodeAttributes& attributes = node->GetAttributes();
+    ORT_ENFORCE(attributes.find("to") != attributes.end());
+    return attributes.at("to").i() == static_cast<int64_t>(data_type);
+  }
+  return false;
+}
+
+static bool IsType(const NodeArg& node_arg, TensorProto_DataType data_type) {
+  return node_arg.TypeAsProto()->tensor_type().elem_type() == data_type;
+}
+
+// InsertCastNodes
+// Insert a new Cast node after each NodeArg in the require_cast vector. The cast node is FLOAT16 if is_fp16 is True
+// and FLOAT otherwise. This funtion fixes the graph edges in addition to inserting the cast nodes.
+static Status InsertCastNodes(Graph& graph,
+                              const std::unordered_set<NodeArg*>& require_cast,
+                              bool is_fp16,
+                              std::deque<onnxruntime::NodeIndex>& removed_nodes) {
+  //Create requirred new Cast nodes.
+  for (NodeArg* node_arg : require_cast) {
+    if (!node_arg->Exists()) {
+      continue;
+    }
+    // data_type is the data type of the Cast output.
+    TensorProto_DataType data_type = is_fp16 ? TensorProto_DataType_FLOAT16 : TensorProto_DataType_FLOAT;
+    TypeProto type_proto;
+    bool is_node_arg_cast_output = IsType(*node_arg, data_type);
+    TensorProto_DataType new_node_arg_data_type = data_type;
+
+    if (is_node_arg_cast_output) {
+      new_node_arg_data_type = (data_type == TensorProto_DataType_FLOAT) ? TensorProto_DataType_FLOAT16 : TensorProto_DataType_FLOAT;
+    }
+    type_proto.mutable_tensor_type()->set_elem_type(new_node_arg_data_type);
+    NodeArg& new_node_arg = graph.GetOrCreateNodeArg(graph.GenerateNodeArgName(node_arg->Name()), &type_proto);
+    // Expect that a NodeArg is not both a graph input as well as a graph output
+    ORT_ENFORCE(!(graph.IsInputsIncludingInitializers(node_arg) && graph.IsOutput(node_arg)));
+    NodeArg& cast_input = !is_node_arg_cast_output ? *node_arg : new_node_arg;
+    NodeArg& cast_output = is_node_arg_cast_output ? *node_arg : new_node_arg;
+    const std::vector<NodeArg*> cast_inputs = {&cast_input};
+    const std::vector<NodeArg*> cast_outputs = {&cast_output};
+    ONNX_NAMESPACE::AttributeProto to_attribute;
+    to_attribute.set_name("to");
+    to_attribute.set_type(ONNX_NAMESPACE::AttributeProto::INT);
+    to_attribute.set_i(static_cast<int64_t>(data_type));
+    NodeAttributes attributes({{"to", to_attribute}});
+
+    Node& cast = graph.AddNode(graph.GenerateNodeName(node_arg->Name() + "_cast"),
+                               "Cast",
+                               "Created a new Cast node",
+                               cast_inputs,
+                               cast_outputs,
+                               &attributes);
+    inserted_node_names.push_back(cast.Name());
+    Node* producer = graph.GetMutableProducerNode(node_arg->Name());
+    std::vector<Node*> consumers = graph.GetMutableConsumerNodes(node_arg->Name());
```

Comment:
non-blocking comment, but this comment applies to the entire file. 
Let's use non-Mutable version of the function whenever we can. 

For example, if following is ok? 
const Node* producer = graph.GetProducerNode(node_arg->Name());